File: leetcode_75/sliding_window/maximum_average_subarray_643.py
from typing import List
class Solution:
    # ----------------------------------------------------------------
    # 63ms
    # 要使用滑動窗
    def findMaxAverage(self, nums: List[int], k: int) -> float:
        temp = sum(nums[0:k])
        max_n = temp

        for n in range(k,len(nums)):
            temp = temp - nums[n-k] + nums[n]
            max_n = max(max_n,temp)
        return max_n/k
    
    # ----------------------------------------------------------------
    # 要避免在loop 中使用sum，他也會掃描選定的區段
    # def findMaxAverage(self, nums: List[int], k: int) -> float:
    #     ll = len(nums)

    #     max_n = float('-inf')
    #     for n in range(ll-k+1):
    #         temp = sum(nums[n:n+k])
    #         if temp>=max_n:
    #             max_n = temp
    #     return max_n/k


aa = Solution()
aa.findMaxAverage(nums = [1,12,-5,-6,50,3], k = 4)

File: leetcode_75/sliding_window/maximum_average_subarray_643.java
class Solution {
    public double findMaxAverage(int[] nums, int k) {
        int temp = 0;
        for (int n = 0; n < k; n++) {
            temp += nums[n];
        }
        int max = temp;

        for (int i = k; i < nums.length; i++) {
            temp = temp - nums[i - k] + nums[i];
            max = Math.max(max, temp);
        }
        return (double) max / k;
    }
}

// 3ms
// (double) max， 要在這邊加上double，讓整體變成小數